funnel/models/event.py

```diff
@@ -0,0 +1,231 @@
+        ticket_type = cls(
+            title=title,
+            proposal_space=space,
+            events=events
+        )
+        db.session.add(ticket_type)
+        return ticket_type
+
+    @classmethod
+    def get_or_create_from(cls, space, title, events=[]):
+        ticket_type = TicketType.get(space, title)
+        if not ticket_type:
+            ticket_type = TicketType.create_from(space, title, events)
+        return ticket_type
+
+
+class Participant(BaseMixin, db.Model):
+    """
+    Model users participating in one or multiple events.
+    """
+    __tablename__ = 'participant'
+
+    fullname = db.Column(db.Unicode(80), nullable=False)
+    #: Unvalidated email address
+    email = db.Column(db.Unicode(80), nullable=False)
+    #: Unvalidated phone number
+    phone = db.Column(db.Unicode(80), nullable=True)
+    #: Unvalidated Twitter id
+    twitter = db.Column(db.Unicode(80), nullable=True)
+    #: Job title
+    job_title = db.Column(db.Unicode(80), nullable=True)
+    #: Company
+    company = db.Column(db.Unicode(80), nullable=True)
+    #: Participant's city
+    city = db.Column(db.Unicode(80), nullable=True)
+    # public key
+    puk = db.Column(db.Unicode(44), nullable=False, default=make_public_key, unique=True)
+    key = db.Column(db.Unicode(44), nullable=False, default=make_private_key, unique=True)
+    badge_printed = db.Column(db.Boolean, default=False, nullable=False)
+    user_id = db.Column(db.Integer, db.ForeignKey('user.id'), nullable=True)
+    user = db.relationship(User, backref=db.backref('participants', cascade='all, delete-orphan'))
+    proposal_space_id = db.Column(db.Integer, db.ForeignKey('proposal_space.id'), nullable=False)
+    proposal_space = db.relationship(ProposalSpace,
+        backref=db.backref('participants', cascade='all, delete-orphan', lazy='dynamic'))
+
+    __table_args__ = (db.UniqueConstraint('proposal_space_id', 'email'),)
+
+    @classmethod
+    def get(cls, space, email):
+        return cls.query.filter_by(proposal_space=space, email=email).first()
+
+    @classmethod
+    def create_from(cls, space, email, fields):
+        participant = cls(
+            fullname=fields.get('fullname'),
+            email=email,
+            phone=fields.get('phone'),
+            twitter=fields.get('twitter'),
+            job_title=fields.get('job_title'),
+            company=fields.get('company'),
+            city=fields.get('city'),
+            events=fields.get('events', []),
+            proposal_space=space
+        )
+        db.session.add(participant)
+        return participant
+
+    @classmethod
+    def get_or_create_from(cls, space, email, fields):
+        participant = Participant.get(space, email)
+        if not participant:
+            participant = Participant.create_from(space, email, fields)
+        return participant
+
+
+class Attendee(BaseMixin, db.Model):
+    """Join model between Participant and Event."""
```

Comment:
`Attendee` should be renamed `EventParticipant` to remove this ambiguity. Our naming convention for join tables is to use the names of the two tables.